compiler/rustc_resolve/src/late.rs

```diff
@@ -1787,6 +1805,32 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
         }
     }
 
+    /// For backwards compatibility, we need to accept the following:
+    ///
+    /// ```
+    /// trait AsStr {
+    ///     fn as_str(&self) -> &str;
+    /// }
+    ///
+    /// fn foo(a: &(dyn AsStr + '_)) -> &str { a.as_str() }
+    /// ```
+    ///
+    /// When we're resolving function parameters, we usually want each elided or
+    /// anonymous lifetime to result in a new elision candidate, but not in this case.
+    #[instrument(level = "trace", skip(self))]
+    fn resolve_anonymous_lifetime_in_trait_object_bounds(&mut self, lifetime: &Lifetime) {
+        debug_assert_eq!(lifetime.ident.name, kw::UnderscoreLifetime);
+        debug_assert!(self.diagnostic_metadata.current_trait_object.is_some());
+
+        if self.in_func_params {
```

Comment:
This is likely to be incorrect in some corner cases. Could you:
- iterate on `self.lifetime_ribs` like `resolve_anonymous_lifetime` does,
- check `LifetimeRibKind::Fresh`?

Is the behaviour restricted to function parameters, or anywhere we have a `Fresh` rib?
What should `impl &(dyn Trait + '_) { /**/ }` mean?

Reply:
`impl &(dyn Trait + '_) { ... }` isn't allowed:

```
error[E0390]: cannot define inherent `impl` for primitive types
  --> /Users/ben/Documents/Rust/rust-lang/rust/src/test/ui/lifetimes/issue-103762-1.rs:42:6
   |
LL | impl &(dyn Trait + '_) {
   |      ^^^^^^^^^^^^^^^^^
   |
   = help: consider using an extension trait instead
   = note: you could also try moving the reference to uses of `dyn Trait` (such as `self`) within the implementation
```

I can believe that this way of doing it is brittle. Can you try to think of any more edge cases? I'm fresh out of ideas at the moment.

I'll try out your strategy of checking for the presence of the rib. Though I think you probably meant `LifetimeRibKind::AnonymousCreateParameter` rather than `LifetimeRibKind::Fresh`?

Reply:
`impl Trait for &(dyn Trait2 + '_) {}` is allowed.

If you want corner cases:
```
fn foo(_: fn(&(dyn Trait + '_)) -> &(dyn Trait + '_)) {}
fn bar(_: [u8; { { fn blah() -> &(dyn Trait + '_) {} }; 5 }]) {}
```

> I'll try out your strategy of checking for the presence of the rib. Though I think you probably meant LifetimeRibKind::AnonymousCreateParameter rather than LifetimeRibKind::Fresh?

Exactly.